smartsim/_core/mli/infrastructure/storage/backbonefeaturestore.py

```diff
@@ -0,0 +1,381 @@
+#    and/or other materials provided with the distribution.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+# AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+# DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+# FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+# DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+# SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+# OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+import enum
+import pickle
+import typing as t
+import uuid
+from collections import defaultdict, deque
+from dataclasses import dataclass
+
+# pylint: disable=import-error
+# isort: off
+import dragon.data.ddict.ddict as dragon_ddict
+
+# isort: on
+
+from smartsim._core.mli.comm.channel.channel import CommChannelBase
+from smartsim._core.mli.infrastructure.storage.dragonfeaturestore import (
+    DragonFeatureStore,
+)
+from smartsim._core.mli.infrastructure.storage.featurestore import ReservedKeys
+from smartsim.error.errors import SmartSimError
+from smartsim.log import get_logger
+
+logger = get_logger(__name__)
+
+
+# todo: did i create an arms race where a developer just grabs the backbone
+# and passes it wherever they need a FeatureStore?
+class BackboneFeatureStore(DragonFeatureStore):
+    """A DragonFeatureStore wrapper with utility methods for accessing shared
+    information stored in the MLI backbone feature store"""
+
+    def __init__(self, storage: "dragon_ddict.DDict") -> None:
+        """Initialize the DragonFeatureStore instance
+
+        :param storage: A distributed dictionary to be used as the underlying
+        storage mechanism of the feature store"""
+        super().__init__(storage)
+        self._reserved_write_enabled = True
+
+    @property
+    def notification_channels(self) -> t.Sequence[str]:
+        """Retrieve descriptors for all registered MLI notification channels
+
+        :returns: the list of descriptors"""
+        if ReservedKeys.MLI_NOTIFY_CONSUMERS in self:
+            stored_consumers = self[ReservedKeys.MLI_NOTIFY_CONSUMERS]
+            return str(stored_consumers).split(",")
+        return []
+
+    @notification_channels.setter
+    def notification_channels(self, values: t.Sequence[str]) -> None:
+        """Set the notification channels to be sent events
+
+        :param values: the list of channel descriptors to save"""
+        self[ReservedKeys.MLI_NOTIFY_CONSUMERS] = ",".join(
+            [str(value) for value in values]
+        )
+
+
+class EventCategory(str, enum.Enum):
```

Comment:
this is planned but needs to happen in the follow up integration PR to avoid losing changes. added todo that will be cleaned up immediately in follow up branch that already exists